File: src/main/java/org/datastructures/thermometernetwork/Sensor.java
package org.datastructures.thermometernetwork;

import java.util.*;

public class Sensor {

    private final String sensorId;
    private final TreeMap<String, TreeMap<String, Double>> dateMap = new TreeMap<>(new DateComparator());

    public Sensor(String sensorId) {
        this.sensorId = sensorId;
    }

    public void addTemperature(String timestamp, double temperature) {
        String date = timestamp.substring(0, 10);
        String time = timestamp.substring(11);
        if (dateMap.containsKey(date)) {
            dateMap.get(date).put(time, temperature);
            return;
        }
        dateMap.put(date, new TreeMap<>(new TimeComparator()));
    }

    //Gets the TreeMap of temps for the supplied date.
    public TreeMap<String, Double> getDayMap(String date) {
        return dateMap.get(date);
    }
    
    //Gets stringified list of temps in map ONLY FOR READABILITY
    public TreeSet<String> getDayTimesAndTemps(TreeMap<String, Double> dayMap) {
    	TreeSet<String> dayTimesAndTemps = new TreeSet<>();
    	dayMap.forEach((key, value)  -> {
    		String combo = key + " : " + value;
    		dayTimesAndTemps.add(combo);
    	});
    	return dayTimesAndTemps;  
    }
    
    public Set<String> getDates(){
    	return dateMap.keySet();
    }

    @Override
    public String toString() {
        return "Sensor: " + sensorId;
    }

}

File: backend/Sensor.py
from collections import OrderedDict
from ordered_set import OrderedSet

class Sensor:
    
    def __init__(self, sensorID):
        self.sensorID = sensorID
        self.dateMap = OrderedDict
    
    def addTemperature(self, timestamp, temperature):
        date = timestamp[0, 10]
        time = timestamp[11:]
        if date in self.dateMap:
            self.dateMap[date] = {time: temperature}
            return
        self.dateMap[date] = {}

    #POTENTIAL BUG: might not be in order, might need to be sorted before returning
    def getTimesAndTemps(self, dayMap):
        timesAndTemps = OrderedSet()

        for time in dayMap:
            combo = time + " : " + dayMap[time]
            timesAndTemps.add(combo)

        return timesAndTemps

    #testing method
    def displayTemperatures(self):
        for date in self.dateMap.keys:
            print(self.getTimesAndTemps(self.dateMap.get(date)))

    def getDates(self):
        return self.dateMap.keys
    
    def getDayMap(self, date):
        return self.dateMap.get(date)
    
    def __str__(self):
        return self.sensorID